draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1216 @@
+:py:meth:`celery.app.task.Task.retry`.
+
+.. code-block:: python
+
+  @task(bind=True)
+  def foo(self):
+    try:
+      # first operation
+    except Exception:
+      self.retry(retry_policy={
+        'max_retries': 10,
+        'interval_start': 0,
+        'interval_step': 5,
+        'interval_max': 120
+      })
+
+    try:
+      first_operation()
+    except Exception:
+      self.retry(retry_policy={
+        'max_retries': 10,
+        'interval_start': 0,
+        'interval_step': 5,
+        'interval_max': 120
+      })
+
+    try:
+      second_operation()
+    except Exception:
+      self.retry(retry_policy={
+        'max_retries': 3,
+        'interval_start': 0,
+        'interval_step': 0.2,
+        'interval_max': 0.2
+      })
+
+Those solutions are unnecessarily verbose. Instead, we could use a with clause
+if all we want to do is retry.
+
+.. code-block:: python
+
+  @task
+  def foo():
+    with retry(max_retries=10, interval_start=0, interval_step=5, interval_max=120):
+      first_operation()
+
+    with retry(max_retries=10, interval_start=0, interval_step=5, interval_max=120):
+      second_operation()
+
+By default messages which cannot be re-published will be stored
+in the :ref:`draft/celery-5-high-level-architecture:messages backlog`.
+
+Implementers may provide other fallbacks such as executing the retried task
+in the same worker or abandoning the task entirely.
+
+Health Checks
+~~~~~~~~~~~~~
+
+Health Checks are used in Celery to verify that a worker is able to
+successfully execute a :ref:`task <draft/celery-5-high-level-architecture:Tasks>`
+or a :ref:`service <draft/celery-5-high-level-architecture:Services>`.
+
+The :ref:`draft/celery-5-high-level-architecture:Scheduler` is responsible
+for scheduling the health checks for execution in each worker after
+each time the configured period of time lapses.
+
+Whenever a health check should be executed the
+:ref:`draft/celery-5-high-level-architecture:Scheduler` instructs the
+:ref:`draft/celery-5-high-level-architecture:Publisher` to send the
+`<health check name>_expired` :term:`Event Message` to each worker's
+:ref:`draft/celery-5-high-level-architecture:Inbox Queue`.
+
+Workers which have tasks subscribed to the event will
+execute all the subscribed tasks in order to determine the state of the
+health check.
+
+Health Checks can handle :term:`Document Messages <Document Message>` as input
+from :ref:`draft/celery-5-high-level-architecture:Ingress Only Data Sources`.
+
+This is useful when you want to respond to an alert from a monitoring system
+or when you want to verify that all incoming data from said source is
+valid at all times before executing the task.
+
+In addition to tasks, Health Checks can also use
+:ref:`draft/celery-5-high-level-architecture:Services` in order to track
+changes in the environment it is running on.
+
+.. admonition:: Example
+
+  We have a task which requires 8GB of memory to complete.
```

Comment:
I understand that this is an example, but is memory consumption monitoring Celery's responsibility? I would think that this concerns mostly optimal message routing, that could possibly be applied by setting constraints on a task level, e.g. `minimum worker memory`, `minimum worker storage space`. What do you think/